// @translate-tools/core/translators/unstable/BingTranslatorPublic/getConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getConfig = void 0;
// This code contains parts from https://github.com/plainheart/bing-translate-api/blob/e5ac48d346096681a758856bd606f147854d07cb/src/index.js
var __awaiter = void 0 && (void 0).__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }
  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }
    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }
    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }
    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};
var __generator = void 0 && (void 0).__generator || function (thisArg, body) {
  var _ = {
      label: 0,
      sent: function () {
        if (t[0] & 1) throw t[1];
        return t[1];
      },
      trys: [],
      ops: []
    },
    f,
    y,
    t,
    g;
  return g = {
    next: verb(0),
    "throw": verb(1),
    "return": verb(2)
  }, typeof Symbol === "function" && (g[Symbol.iterator] = function () {
    return this;
  }), g;
  function verb(n) {
    return function (v) {
      return step([n, v]);
    };
  }
  function step(op) {
    if (f) throw new TypeError("Generator is already executing.");
    while (g && (g = 0, op[0] && (_ = 0)), _) try {
      if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
      if (y = 0, t) op = [op[0] & 2, t.value];
      switch (op[0]) {
        case 0:
        case 1:
          t = op;
          break;
        case 4:
          _.label++;
          return {
            value: op[1],
            done: false
          };
        case 5:
          _.label++;
          y = op[1];
          op = [0];
          continue;
        case 7:
          op = _.ops.pop();
          _.trys.pop();
          continue;
        default:
          if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
            _ = 0;
            continue;
          }
          if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
            _.label = op[1];
            break;
          }
          if (op[0] === 6 && _.label < t[1]) {
            _.label = t[1];
            t = op;
            break;
          }
          if (t && _.label < t[2]) {
            _.label = t[2];
            _.ops.push(op);
            break;
          }
          if (t[2]) _.ops.pop();
          _.trys.pop();
          continue;
      }
      op = body.call(thisArg, _);
    } catch (e) {
      op = [6, e];
      y = 0;
    } finally {
      f = t = 0;
    }
    if (op[0] & 5) throw op[1];
    return {
      value: op[0] ? op[1] : void 0,
      done: true
    };
  }
};
var __read = void 0 && (void 0).__read || function (o, n) {
  var m = typeof Symbol === "function" && o[Symbol.iterator];
  if (!m) return o;
  var i = m.call(o),
    r,
    ar = [],
    e;
  try {
    while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
  } catch (error) {
    e = {
      error: error
    };
  } finally {
    try {
      if (r && !r.done && (m = i["return"])) m.call(i);
    } finally {
      if (e) throw e.error;
    }
  }
  return ar;
};
var fetchConfig = function () {
  return __awaiter(void 0, void 0, void 0, function () {
    var pageText, IIGSearch, IIG, IIDSearch, IID, TokenParamsSearch, _a, key, token, lifetime;
    return __generator(this, function (_b) {
      switch (_b.label) {
        case 0:
          return [4 /*yield*/, fetch('https://www.bing.com/translator').then(function (r) {
            return r.text();
          })];
        case 1:
          pageText = _b.sent();
          IIGSearch = pageText.match(/IG:"([^"]+)"/);
          if (IIGSearch === null) {
            throw new Error("Can't find IIG");
          }
          IIG = IIGSearch[1];
          IIDSearch = pageText.match(/data-iid="([^"]+)"/);
          if (IIDSearch === null) {
            throw new Error("Can't find IID");
          }
          IID = IIDSearch[1];
          TokenParamsSearch = pageText.match(/params_RichTranslateHelper\s?=\s?([^\]]+\])/);
          if (TokenParamsSearch === null) {
            throw new Error("Can't find token params");
          }
          _a = __read(JSON.parse(TokenParamsSearch[1]), 3), key = _a[0], token = _a[1], lifetime = _a[2];
          if (typeof key !== 'number') {
            throw new Error('Invalid key type');
          }
          if (typeof token !== 'string') {
            throw new Error('Invalid token type');
          }
          if (typeof lifetime !== 'number') {
            throw new Error('Invalid token lifetime type');
          }
          return [2 /*return*/, {
            IID: IID,
            IIG: IIG,
            key: key,
            token: token,
            lifetime: lifetime
          }];
      }
    });
  });
};
var lastConfig = null;
function isTokenExpired() {
  if (lastConfig === null) {
    return true;
  }
  var tokenTs = lastConfig.key,
    tokenExpiryInterval = lastConfig.lifetime;
  return Date.now() - tokenTs > tokenExpiryInterval;
}
var getConfig = function () {
  return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          if (!isTokenExpired()) return [3 /*break*/, 2];
          return [4 /*yield*/, fetchConfig()];
        case 1:
          lastConfig = _a.sent();
          _a.label = 2;
        case 2:
          return [2 /*return*/, lastConfig];
      }
    });
  });
};
exports.getConfig = getConfig;
